### PR TITLE
add csrf exception for token endpoint

### PR DESCRIPTION
**Pull request by mhenschke-atl**
Based on the `BitbucketWebookCrumbExclusion`. Need to spin up a war with csrf filtering on locally

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/TokenCrumbExclusion.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+
+@Extension
+public class TokenCrumbExclusion extends CrumbExclusion {
+
+    static final String EXCLUSION_PATH = "/bitbucket/oauth/";
```
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/TokenCrumbExclusion.java:17]: I wanted to define this in terms of constants, but the path is generated by stapler, so I've hard-coded it here

**Review by khughes-atlassian — APPROVED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/TokenCrumbExclusion.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+import static org.apache.commons.lang3.StringUtils.isEmpty;
+
+@Extension
+public class TokenCrumbExclusion extends CrumbExclusion {
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/TokenCrumbExclusion.java:15]: We already have a `com.atlassian.bitbucket.jenkins.internal.applink.oauth.serviceprovider.rest.OauthCrumbExclusion` maybe we could add this check to it? It seems like this is just an additional case that the `OauthCrumbExclusion` should have handled?

Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/TokenCrumbExclusion.java:15]: I was basing this off of the design for the webhook exception, and thought it was "one exception per case". That said, they are very similar- happy to migrate the code into that class to keep things simple :)

Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/TokenCrumbExclusion.java:15]: I think the resources are per "task" so anything for Oauth should be in the oauth filter.

Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/TokenCrumbExclusion.java:15]: @mhenschke-atl we don't need this file any more, right? nor the test file for it.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/OAuthRequestUtils.java [hunk 3]
```diff
@@ -36,7 +38,18 @@ private OAuthRequestUtils() {
      * @return true if the request is an OAuth request.
      */
     public static boolean isOAuthAccessAttempt(HttpServletRequest request) {
-        return is3LOAuthAccessAttempt(request) || is2LOAuthAccessAttempt(request);
+        return is3LOAuthAccessAttempt(request) || is2LOAuthAccessAttempt(request) || isOauthTokenRequest(request) || isOauthTokenRequest(request);
```
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/OAuthRequestUtils.java:41]: this is also used by the `OAuth1aRequestFilter` maybe we should ask for a comment from @goravjoshi  to confirm that this won't have any unintended consequences for the general oauth flow?

**Review by dkjellin — CHANGES_REQUESTED** [review]
I think we should remove the `TokenCrumbExclusion` class as it is no longer needed, plus I think we need to be certain that the inclusion of the token path won't have any unintended side-effects for a general oauth flow.

Requesting changes mainly to get the new-and-no-longer-needed files removed.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/TokenCrumbExclusion.java (hunk 2 above)
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/rest/TokenCrumbExclusion.java:15]: We don't :facepalm: I didn't push the delete. Sorry!

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/OAuthRequestUtils.java (hunk 3 above)
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/OAuthRequestUtils.java:41]: That's a really good point. I think it's better we just separate it into a different method and call both from the crumb exception. The other methods explicitly exclude request and access tokens

Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/applink/oauth/serviceprovider/OAuthRequestUtils.java:41]: Also the conditional is being tested twice... something weird happened with this commit...

**Review by dkjellin — APPROVED** [review]

**Pull request merged**